tests/sbd-testbed.c

```diff
@@ -0,0 +1,746 @@
+{
+    int rv = -1;
+    va_list ap;
+    int i;
+
+    init();
+
+    va_start(ap, request);
+    switch (request) {
+        case BLKSSZGET:
+            for (i=0; i < 3; i++) {
+                if (sbd_device_fd[i] == fd) {
+                    rv = 0;
+                    *(va_arg(ap, int *)) = 512;
+                    break;
+                }
+                if (i == 2) {
+                    rv = orig_ioctl(fd, request, va_arg(ap, int *));
+                }
+            }
+            break;
+        case WDIOC_SETTIMEOUT:
+            if (fd == watchdog_device_fd) {
+                watchdog_timeout = *va_arg(ap, int *);
+
+                watchdog_arm();
+                rv = 0;
+                break;
+            }
+            rv = orig_ioctl(fd, request, va_arg(ap, int *));
+            break;
+        case WDIOC_SETOPTIONS:
+            if (fd == watchdog_device_fd) {
+                int flags = *va_arg(ap, int *);
+
+                if (flags & WDIOS_DISABLECARD) {
+                    watchdog_disarm();
+                }
+                rv = 0;
+                break;
+            }
+            rv = orig_ioctl(fd, request, va_arg(ap, int *));
+            break;
+        case WDIOC_GETSUPPORT:
+            rv = orig_ioctl(fd, request, va_arg(ap, struct watchdog_info *));
+            break;
+        default:
+            fprintf(log_fp, "ioctl using unknown request = 0x%08lx", request);
+            rv = orig_ioctl(fd, request, va_arg(ap, void *));
+    }
+    va_end(ap);
+
+    return rv;
+}
+
+int
+close(int fd)
+{
+    int i;
+
+    init();
+
+    if (fd == watchdog_device_fd) {
+        watchdog_device_fd = -1;
+    } else {
+        for (i = 0; i < 3; i++) {
+            if (sbd_device_fd[i] == fd) {
+                sbd_device_fd[i] = -1;
+                break;
+            }
+        }
+    }
+    return orig_close(fd);
+}
+
+// ***** end - handling of watchdog & block-devices ****
+
+// ***** handling of sysrq, sysrq-trigger & reboot ****
+
+FILE *
+fopen(const char *pathname, const char *mode)
+{
+    int is_sysrq = 0;
+    int is_sysrq_trigger = 0;
+    FILE *fp;
+
+    init();
+
+    if ((strcmp("/proc/sys/kernel/sysrq", pathname) == 0) &&
+        strcmp("w", mode)) {
+        pathname = "/dev/null";
+        is_sysrq = 1;
+    } else if (strcmp("/proc/sysrq-trigger", pathname) == 0) {
+        pathname = "/dev/null";
+        is_sysrq_trigger = 1;
+    }
+    fp = orig_fopen(pathname, mode);
+    if (is_sysrq) {
+        sysrq_fp = fp;
+    } else if (is_sysrq_trigger) {
+        sysrq_trigger_fp = fp;
+    }
+    return fp;
+}
+
+int
+fclose(FILE *fp)
+{
+    init();
+
+    if (fp == sysrq_fp) {
+        sysrq_fp = NULL;
+    } else if (fp == sysrq_trigger_fp) {
+        sysrq_trigger_fp = NULL;
+    }
+    return orig_fclose(fp);
+}
+
+#if __USE_FORTIFY_LEVEL > 1
+
+int
+__fprintf_chk(FILE *stream, int flag, const char *format, ...)
+{
+    va_list ap;
+    int rv;
+
+    init();
+    va_start (ap, format);
```

Comment:
Not a big deal, but you could avoid duplication by functionizing the part of after this, taking ap as argument

Reply:
good point ... but even simpler ... that should already do the trick:
#if __USE_FORTIFY_LEVEL > 1
int __fprintf_chk(FILE *stream, int flag, const char *format, ...)
#else
int fprintf(FILE *stream, const char *format, ...)
#endif